### PR TITLE
##3699938 Implement comment vote and update

### PR DESCRIPTION
**Pull request by andela-oowonikoko**
### What does this PR do ?

- Implements comment vote and comment update
- Update comment data to include `ownerName` and `ownerImageUrl`
- Fixes some pep8 errors

### Description of Task to be completed ? 

- create and test the endpoint /comments/<string:comment_id> to update a comment
- create and test the endpoint /comment/<string:comment_id>/votes to register and retract votes
- implements and tests logic to add the hasVoted field to comments retrieved
- implements and tests logic to add voteCount field to comments retrieved
- adds pagination to comments
- update relationship between user and comment model
- modify get contributions endpoint to post to user Table
- update the comment data sent to the frontend and tests

### How should this be manually tested?

- At the root of the project directory run the following command:
``` bash
  pytest -v or pytest
```
What are the relevant pivotal tracker stories?

#150037772, #150400838, #150400721, #150400956 and #150522054

### CODE REVIEW

File: api/views/contributions.py [hunk 1]
```diff
@@ -273,11 +280,51 @@ def delete(self, comment_id):
...
 
+    @token_required
+    def put(self, comment_id):
+        comment = Comment.query.get(comment_id)
+        if not comment:
+            return {"message":
+                    "The specified comment was not found."}, 404
+
+        new_content = request.json.get("message")
+        if not new_content or not new_content.strip():
```
[the remark this answers is not in the document]
Comment (georgreen) [reply api/views/contributions.py]: @LarryWachira I don't think so, if new_content is None, that will throw an error but with the first check this code is safe.

File: api/views/contributions.py [hunk 2]
```diff
@@ -273,11 +280,51 @@ def delete(self, comment_id):
...
+            return {"message":
+                    "Specify the message for new comment."}, 400
+        elif new_content == comment.message:
```
[the remark this answers is not in the document]
Comment (georgreen) [reply api/views/contributions.py]: @LarryWachira Yes, we decided to disallow users from updating comments with the same message. e.g updating a comment, "that was cool" to "that was cool".

File: api/views/contributions.py [hunk 3]
```diff
@@ -113,20 +121,19 @@ def post(self):
...
+                    tag.serialize() for tag in tagged_contribution.tags]
+
                 # Save the new contribution to the database
                 tagged_contribution.save()
 
             # add contribution_type field
-            contribution_serialized['contribution_type'] = new_contribution.contribution_type.name
-            
+            contribution_type = new_contribution.contribution_type.name
+            contribution_serialized['contribution_type'] = contribution_type
```
[the remark this answers is not in the document]
Comment (andela-oowonikoko) [reply api/views/contributions.py]: Fixed

File: api/views/contributions.py [hunk 4]
```diff
@@ -218,33 +226,33 @@ def get(self, contrib_id):
...
-        # fetch contribution from data base
         contribution = Contribution.query.get(contrib_id)
         parent_id = None
-        # check if contributiion exists
+
         if not contribution:
             return {"message": "contribution does not exist"}, 404
-        #check if request is a json
+
         if not request.json:
```
[the remark this answers is not in the document]
Comment (andela-oowonikoko) [reply api/views/contributions.py]: Fixed!!!

File: api/views/contributions.py [hunk 5]
```diff
@@ -218,33 +226,33 @@ def get(self, contrib_id):
...
             return {"message": "Request must be a valid JSON"}, 400
+
         if request.get_json():
```
[the remark this answers is not in the document]
Comment (andela-oowonikoko) [reply api/views/contributions.py]: Fixed

File: api/views/contributions.py [hunk 6]
```diff
@@ -218,33 +226,33 @@ def get(self, contrib_id):
...
                     return {"message": "parent comment does not exist"}, 400
                 if parent_comment.parent_id:
-                    return {"message": "A comment reply cannot have a reply"}, 400
+                    return {"message":
+                            "A comment reply cannot have a reply"}, 400
                 if parent_comment:
                     parent_id = parent_comment.id
                     contrib_id = None
+
             generated_id = uuid4().hex
```
[the remark this answers is not in the document]
Comment (andela-oowonikoko) [reply api/views/contributions.py]: Fixed

File: api/views/contributions.py [hunk 7]
```diff
@@ -306,10 +341,50 @@ def delete(self, comment_id):
...
+        new_content = request.json.get("message")
+        if not new_content or not new_content.strip():
+            return {"message":
+                    "Specify the message for new comment."}, 400
+        elif new_content == comment.message:
+            return {"message":
+                    "Can not update comment with the same content."}, 409
+        elif g.current_user.user_id == comment.user_id:
+            comment.message = new_content
+            db.session.add(comment)
```
Comment (andela-ojoloko) [inline api/views/contributions.py]: I believe this can be replaced with `comment.save`

Comment (andela-oowonikoko) [reply api/views/contributions.py]: Fixed

**Comment by andela-ojoloko** [comment]
LGTM 👍

File: api/views/contributions.py [hunk 8]
```diff
@@ -236,65 +245,91 @@ def get(self, contrib_id):
...
+        if _comments.has_prev:
+            previous_url = url_for(request.endpoint, contrib_id=contrib_id,
+                                   limit=limit, page=page-1, _external=True
+                                  )
+
+        return jsonify(dict(data={"comments": comments},
                             commentCount=len(comments),
-                            success="You have succesfully retrieved "
-                                    "the comments"))
+                            next_url=next_url,
```
Comment (LarryWachira) [inline api/views/contributions.py]: camelCase

Comment (LarryWachira) [reply api/views/contributions.py]: Also, responses do not comply with what we agreed. (JSEND)

Comment (andela-oowonikoko) [reply api/views/contributions.py]: Fixed

**Comment by LarryWachira** [comment]
LGTM

**Comment by danielwangai** [comment]
LGTM

**Pull request merged**